### ISSUE TITLE
Gradient type feedback

### ISSUE THREAD

**Issue by c1rrus** [design-tokens/community-group#101]
Is the specification for gradients that is being introduced by PR #86 fit for purpose? Does it need to also specify the type of gradient (.e.g linear, radial, concial, etc.)?

Please share your feedback, thoughts and ideas in this issue.

**Comment by drwpow** [comment]
> Does it need to also specify the type of gradient (.e.g linear, radial, concial, etc.)? 

I do really like the proposal here, and I like the simplicity of having the gradient stops be universal, where you decide linear, radial, conical when generating code for a specific component (which in one part makes sense because a rotated gradient is still the same gradient).

But there also seems like an opportunity here to make `linear-gradient,` `radial-gradient`, and `conic-gradient` their own types to fill in the missing data (which exist as a concept both in CSS and Figma). Overall I fear that people may just abuse `metadata` on gradients here and put all the missing parts in there, at which point distinct types would be more useful. I could go either way—one gradient type with stops only, or distinct gradient types with complete properties.

**Comment by TravisSpomer** [comment]
Omitting details like the type of gradient doesn't make much sense to me. A reasonable expectation is that for any given token, a design tool has everything it needs to display that gradient (like a Figma style), and a dev tool has everything it needs to turn that token definition into code. The current gradient spec doesn't work for either case. As it stands, I'm not really sure how I could use a gradient token for anything at all, without encoding all of the missing information as extensions.

At absolute minimum, I think that a gradient token needs to also include:

* The type of gradient
* The start and end points (0-1)

It would also be unfortunate if gradient stops were limited to specifying stops as a number 0-1. Other systems also allow stops to be defined as absolute units. For example, you might specify a gradient as going from black at 0, to transparent at 10px, to transparent at 1—that would get you a "poor man's shadow" that was always 10px wide no matter the size of the object you paint with it. That's admittedly a less common case, but it's something that I'm using today to support web and WinUI, and to continue supporting it I'd need to add an extension property like `"actuallyTheseAreTheRealGradientStops": [ 0, "10px", 1 ]`.

**Comment by solemone** [comment]
Hello everyone, 

I have a question about the number of gradients. How about multiple gradients? Is the value then an array with arrays? Like like this:

```
{
  "blue-to-red": {
    "$type": "gradient",
    "$value": [
      [
        {
          "color": "rgba(0, 0, 255, 0.5)",
          "position": 0
        },
        {
          "color": "rgba(0, 0, 255, 1)",
          "position": 1
        }
      ],
      [
        {
          "color": "rgba(255, 0, 0, 1)",
          "position": 0
        },
        {
          "color": "rgba(255, 0, 0, 0.5)",
          "position": 1
        }
      ]
    ]
  }
}
```

Or is this a different type like »multifill«? 
And what about when gradients are to be mixed with solid color? Could this be something like this:

```
{
  "blue-and-red": {
    "$type": "multifill",
    "$value": [
      {
        "$type": "color",
        "$value": "rgba(0, 0, 255, 0.5)",
      },
      {
        "$type": "gradient",
        "$value": [
          {
            "color": "rgba(255, 0, 0, 0.5)",
            "position": 0
          },
          {
            "color": "rgba(255, 0, 0, 1)",
            "position": 1
          }
        ]
      }
    ]
  }
}
```
Thanks for all the work!

**Comment by NateBaldwinDesign** [comment]
I agree supporting other gradient types is necessary, but I am of the opinion it doesn't need to be a different token type itself.

I would like a few things here:
First, it would be nice to have a default inferred from a list of values. Ie, `values: [#ff0000, #0000ff]` will automatically infer placement to be [0,1]. Reason being, it may be unnecessary, bloat, or redundant based on certain use cases. My next ask will shed more light.

Secondly, I would like the option to just pass an array of values. I go at great lengths to make perceptually balanced gradients, and given technological restrictions, most of the time an array of many values (spaced equidistant in the range) is the best way to replicate the gradient. For instance, if I have a gradient with 20 unique colors, it would be tedious to write all the intervals in the proposed format.

Thirdly, it would be great to include an option for interpolation color space. Technology is working to catch up on this notion and certain design tools support it. For instance, defaulting to rgb interpolation is ok, however I may want to use OKLCH for a multi hue gradient. Interpolation color space options can reduce the number of individual color values needed to replicate a perceptually balanced gradient (within these examples)

**Comment by putchom** [comment]
I think `$value` needs an angle property because Figma, CSS, etc. allow angles to be specified for gradients.

like this:

```json
{
  "blue-to-red": {
    "$type": "gradient",
    "$value": {
      "angle": 45,
      "color": [
        {
          "color": "#0000ff",
          "position": 0
        },
        {
          "color": "#ff0000",
          "position": 1
        }
      ]
    }
  }
}
```

**Comment by PavelLaptev** [comment]
Agreed with @drwpow `gradient` tokens is a very abstract entity. So there should be a definition:
```
GradientToken {
  $type: "gradient";
  $value: {
    type: "linear" | "radial";
    angle: number;
    stops: {
      position: number;
      color: ColorToken | string;
    }[];
  };
}
```

**Comment by PavelLaptev** [comment]
For "radial" gradients, there also should be a size property, e.g.:
```
GradientToken {
  $type: "gradient";
  $value: {
    type: "linear" | "radial" | "angular" | "conic";
    angle: number;
    size: "30% 50%"
    stops: {
      position: number;
      color: ColorToken | string;
    }[];
  };
}
```

**Comment by jorenbroekema** [comment]
I wanted to add to this discussion that the gradient types I've heard here are mostly linear by nature, even though the shapes can be defined as "linear", "conicol", "angular", "radial", "diamond", whatever, the color stops are a single dimensional array.

There are also types of gradient which I see used more and more in modern designs, which cannot be defined by a string enum to describe statically what the shape of the gradient looks like:
- Gradient meshes
- Freeform gradients
E.g. in Adobe Illustrator, example video: https://www.youtube.com/watch?v=fhB5tTTLqDU
Especially for freeform gradients, I can imagine huge platform-specific differences in how they are rendered, which is already a factor for simple gradients.

I think it would be smart to at least distinguish between these two types of gradients, once that have a standardized "shape" and those that do not. And at least not jam those two into the same token type 😅 .

For the standardized shapes gradients, the main thing I'm missing:
- color stops are defined on a 1 dimensional plane, so the angle of the gradient is not included, I think it makes sense that gradient stops have X Y and maybe even Z coordinates for 3D gradients, which could be used in 3D rendering software where 3D color interpolation is actually pretty common. This also means we don't need a separate "angle" property. I think this aligns more with how design tools currently let you work with gradients (Illustrator gradient tool, figma gradient paint)
- easing, e.g. a percentage or number between 0 and 1 that describes the relationship between 2 stops in how strong one is compared to the other, I wonder whether that should be defined inside the stops or in a sibling property that is indexed to the stops array somehow.

**Comment by lukasoppermann** [comment]
## Missing specification for stop amount

Hey, the specs don't say anything about the number of stops needed.
To me it would make sense if at least 2 stops would be required. But maybe there is a case for 1 stop?

I think this should be mentioned in the docs.

**Comment by Shrinks99** [comment]
@lukasoppermann 1 stop is [officially supported with css-images-4](https://github.com/w3c/csswg-drafts/issues/10092) and has use in CSS with `mask:` https://css-tip.com/border-gradient/
